### PR TITLE
[llvm-mca] Abort on parse error without -skip-unsupported-instructions

### PR DESCRIPTION
**Pull request by peterwaller-arm**
[llvm-mca] Abort on parse error without -skip-unsupported-instructions

Prior to this patch, llvm-mca would continue executing after parse
errors. These errors can lead to some confusion since some analysis
results are printed on the standard output, and they're printed after
the errors, which could otherwise be easy to miss.

However it is still useful to be able to continue analysis after errors;
so extend the recently added -skip-unsupported-instructions to support
this.

Two tests which have parse errors for some of the 'RUN' branches are
updated to use -skip-unsupported-instructions so they can remain as-is.

Add a description of -skip-unsupported-instructions to the llvm-mca
command guide, and add it to the llvm-mca --help output:

```
  --skip-unsupported-instructions=<value> - Force analysis to continue in the presence of unsupported instructions
    =none                                 -   Exit with an error when an instruction is unsupported for any reason (default)
    =lack-sched                           -   Skip instructions on input which lack scheduling information
    =parse-failure                        -   Skip lines on the input which fail to parse for any reason
    =any                                  -   Skip instructions or lines on input which are unsupported for any reason
```

Tests within this patch are intended to cover each of the cases.

Reason        | Flag | Comment
--------------|------|-------
none          | none | Usual case, existing test suite
lack-sched    | none | Advises user to use -skip-unsupported-instructions=lack-sched, tested in llvm/test/tools/llvm-mca/X86/BtVer2/unsupported-instruction.s
parse-failure | none | Advises user to use -skip-unsupported-instructions=parse-failure, tested in llvm/test/tools/llvm-mca/bad-input.s
any           | none | (N/A, covered above)
lack-sched    | any  | Continues, prints warnings, tested in llvm/test/tools/llvm-mca/X86/BtVer2/unsupported-instruction.s
parse-failure | any  | Continues, prints errors, tested in llvm/test/tools/llvm-mca/bad-input.s
lack-sched    | parse-failure | Advises user to use -skip-unsupported-instructions=lack-sched, tested in llvm/test/tools/llvm-mca/X86/BtVer2/unsupported-instruction.s
parse-failure | lack-sched    | Advises user to use -skip-unsupported-instructions=parse-failure, tested in llvm/test/tools/llvm-mca/bad-input.s
none          | * | This would be any test case with skip-unsupported-instructions, coverage added in llvm/test/tools/llvm-mca/X86/BtVer2/simple-test.s

### CODE REVIEW

**Comment by adibiagio** [comment]
Could you please add a brief description of flag `-skip-unsupported-instructions` to the llvm-mca command guide?

**Comment by michaelmaitland** [comment]
I understand skipping instructions that lack scheduling information but do we really mean to allow instructions that don’t parse?

**Comment by peterwaller-arm** [comment]
> I understand skipping instructions that lack scheduling information but do we really mean to allow instructions that don’t parse?

One example use case of this is in the exynos test modified by this PR. That one has instructions in the test case which are not supported. Previously, MCA would happily continue to report something anyway. Now that I've changed this not to be default, it's necessary to be able to skip them again to make the test work.

Generally I think this is useful to have. You could find yourself with an assembly you want to analyse which has a rare instruction in it which hasn't yet had LLVM support added, for example if you're using an old version of LLVM at some point in the future and you want to see what MCA would have said if it were not for the unsupported instruction.

**Comment by peterwaller-arm** [comment]
I'm happy having a separate flag if that seems beneficial (commence bikeshed: `-skip-parse-errors`? Naming seems tricky given that the parser emits errors if you use instructions which are not supported with the current feature flags, for example).

However, I'm unclear on how beneficial it is to split this case up to a user. A user would still be able to distinguish the two according to the stderr errors (and warnings, notes) that are printed.

Do we really want to force a user to distinguish the reasons for skipping the flags on input? Essentially, if I was faced with a failure due to the presence of some instructions on the input, I would want to reach for the one obvious flag to proceed, rather than to have to figure out which of the N > 1 flags it is.

Another possibility would be to have -skip-unsupported-instructions not be just a bool but a flag that takes multiple options, ranging from `-skip-unsupported-instructions=no-sched-info`, `-skip-unsupported-instructions=parse-failure` to `-skip-unsupported-instructions=all`.

**Comment by michaelmaitland** [comment]
> Maybe we should add another flag for this particular use case?

> Another possibility would be to have -skip-unsupported-instructions

Either sounds sounds like a good solution. I think there are two distinct cases here:
  1. Cant parse to MCInstr
  2. Cant parse from MCInstr to MCAInstr

We should be careful to not conflate them.

>  That one has instructions in the test case which are not supported (edit: for some variations of -mcpu tested)

I'm more worried about passing something totally incorrect, like passing not enough registers, or incorrect types of registers, or even mnemonics that don't exist. I also wonder if it would be hygienic to have a test case where you run it on multiple mcpu but one CPU cannot accept that program.

**Comment by peterwaller-arm** [comment]
OK, I propose to:

* make llvm-mca report an error and exit by default
* make `-skip-unsupported-instructions` take `no-sched-info`, `parse-error` and `all`.
* `parse-error` would skip the case `Parser->Run(false)` returning true (indicating error).
* Add a test case to cover your suggestion 'I also wonder if it would be hygienic to have a test case where you run it on multiple mcpu but one CPU cannot accept that program.'

I just want to clarify what you're asking there; it's that this should test the thing that the exynos case currently hits: we have an instruction where some mcpu can accept the instruction, and others can't, and all have `-skip-unsupported-instructions=parse-error` set, and it checks that the total cycle count is as expected? I would derive such a test case from the existing exynos case.

**Comment by michaelmaitland** [comment]
To ignore parse errors is to accept programs that are not well formed. Do we really want to support this?

The proposal doesn't account for this concern:

> I'm more worried about passing something totally incorrect, like passing not enough registers, or incorrect types of registers, or even mnemonics that don't exist.

What do you think about this proposal:
* the test case should use a different instruction sequence that both CPU can support, or the test case should be split into two cases, one for each CPU.
* -skip-unsupported-instructions only skips instructions that cannot be converted from MCInst to mca::Instruction

**Comment by peterwaller-arm** [comment]
I accept your points. The root of my thinking was to preserve backwards compatibility, but I think there are potential reasons to want this.

> To ignore parse errors is to accept programs that are not well formed. Do we really want to support this? Is it useful to have this behavior? What use cases does this have? You wouldn't give a CPU a program with instructions it couldn't support and expect it to work correctly. The compiler wouldn't generate that code for that CPU either.

* llvm-mca today accepts bad input and prints analysis outputs.
* It seems likely that was unintentional, but if anyone is relying on this, accidentally or otherwise, it would be nice to 'give them an out' if they need it.
* tomorrow's new instructions are today's malformed inputs. Yes, one hopes LLVM will gain support for those new instructions, and that a user will be able to upgrade their llvm-mca to gain support of those, but I still think it is useful to have an alternative route if either of these conditions are false for some time. 
* Imagine a case where someone is trying to bisect a behaviour change in llvm-mca, or study mca's output as a function of time for a given (large) set of inputs. They could find themselves running an old llvm-mca against new code. If that use case involved throwing 'lots of code at llvm-mca', it may be useful to ignore instructions which don't parse at older revisions of llvm.

MCA and LLVM's MC parser seems to already handle bad input quite well, and assembler syntax lends itself is simple enough for this. There aren't lots of complex ambiguities that are introduced by ignoring a line, that I can think of, correct me if I'm wrong.

The analysis outputs may be compromised to some degree for bad inputs, but so long as the user is being explicit about allowing it and being warned that the analysis will be less accurate, I don't see the harm in supporting this case.

> the test case should use a different instruction sequence that both CPU can support, or the test case should be split into two cases, one for each CPU.

When I first saw this test case was running with errors behind the scenes I thought this was unfortunate, but later I thought the [test case](
https://github.com/llvm/llvm-project/blob/6b57520107755e8721b61cd4fcd99461be29f5c9/llvm/test/tools/llvm-mca/AArch64/Exynos/float-divide-multiply.s) as it stands is pretty neat for what it's worth. It's autogenerated and has all the CPU generations in one file, and the generation lacking support for the instruction neatly omits the unsupported instructions from the output.

I don't feel the case above is very strong, but I'd like to pursue this unless strong objections are raised.

**Comment by adibiagio** [comment]
> * llvm-mca today accepts bad input and prints analysis outputs.
> * It seems likely that was unintentional, but if anyone is relying on this, accidentally or otherwise, it would be nice to 'give them an out' if they need it.

I can tell you that it was not intentional. In case of parse errors, it leads to wrong/misleading perf reports. I may be wrong, but I don't think that this is what most users want. In general, I don't think that people should rely on this. 

As I mentioned in my previous message, we could add a separate option to specifically ignore parse errors.
When erroring out, we could print a remark message that says "Please use flag -XYZ` to explicitly ignore parse errors.`.
That way, people will know how to fix it if they get confused by this behaviour change.

Just my opinion.

**Comment by peterwaller-arm** [comment]
Thanks for coming around and allowing my side :). I am also motivated to keep things simple.

Ideally the decision isn't permanent; I hope we make a design now which gives a way forward where we can extend things for more specificity if new information comes to light.

So that said I'm swinging back to this proposal, which is my current intent.

> https://github.com/llvm/llvm-project/pull/90474#issuecomment-2083198408
>
> OK, I propose to:
> 
> * make llvm-mca report an error and exit by default
> * make -skip-unsupported-instructions take no-sched-info, parse-error and all.
> * parse-error would skip the case Parser->Run(false) returning true (indicating error).


That is:

* In aid of keeping things simple we have one flag, `-skip-unsupported-instructions`, and it takes `no-sched-info`, `parse-error` and `all`. If it becomes desirable to add more distinction it should be possible to add other values to this option.
* The flag describes what it is doing, which is skipping over instructions on the input.

I'll hold off implementing this for a day or two in case there are more comments refining this.

**Comment by adibiagio** [comment]
> Thanks for coming around and allowing my side :). I am also motivated to keep things simple.
> 
> Ideally the decision isn't permanent; I hope we make a design now which gives a way forward where we can extend things for more specificity if new information comes to light.
> 
> So that said I'm swinging back to this proposal, which is my current intent.
> 
> > [#90474 (comment)](https://github.com/llvm/llvm-project/pull/90474#issuecomment-2083198408)
> > OK, I propose to:
> > 
> > * make llvm-mca report an error and exit by default
> > * make -skip-unsupported-instructions take no-sched-info, parse-error and all.
> > * parse-error would skip the case Parser->Run(false) returning true (indicating error).
> 
> That is:
> 
> * In aid of keeping things simple we have one flag, `-skip-unsupported-instructions`, and it takes `no-sched-info`, `parse-error` and `all`. If it becomes desirable to add more distinction it should be possible to add other values to this option.
> * The flag describes what it is doing, which is skipping over instructions on the input.
> 
> I'll hold off implementing this for a day or two in case there are more comments refining this.

Sounds reasonable to me.
Thanks

**Comment by holland11** [comment]
Just read through the whole thread and I agree with the consensus up to this point. 

> * In aid of keeping things simple we have one flag, `-skip-unsupported-instructions`, and it takes `no-sched-info`, `parse-error` and `all`. If it becomes desirable to add more distinction it should be possible to add other values to this option.
> 
> * The flag describes what it is doing, which is skipping over instructions on the input.

Implementing it like this seems completely reasonable to me. Thanks for the effort.

**Review by adibiagio — APPROVED** [review]
LGTM.

Thanks!

**Review by michaelmaitland — APPROVED** [review]
LGTM. Thanks for addressing this.

**Comment by peterwaller-arm** [comment]
Thanks for helping find consensus. I found one more issue which I've addressed: there were some thumb and other target=ARM instructions which weren't parsing. Rather than skip them I've fixed them in situ. I'll merge when I return to work on Tuesday next week assuming that CI is green or there are only trivial things to fix.

File: llvm/test/tools/llvm-mca/bad-input.s
Comment (peterwaller-arm) [inline llvm/test/tools/llvm-mca/bad-input.s:5]: Reference: https://lab.llvm.org/buildbot/#/builders/60/builds/17086

I've just become aware that this appears to be failing on builders with messages like:

```
not c:\buildbot\as-builder-1\x-armv7l\build\bin\llvm-mca.exe C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s -o /dev/null 2>&1 | c:\buildbot\as-builder-1\x-armv7l\build\bin\filecheck.exe --check-prefixes=CHECK-ALL,CHECK C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s
# executed command: not 'c:\buildbot\as-builder-1\x-armv7l\build\bin\llvm-mca.exe' 'C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s' -o /dev/null
# executed command: 'c:\buildbot\as-builder-1\x-armv7l\build\bin\filecheck.exe' --check-prefixes=CHECK-ALL,CHECK 'C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s'
# .---command stderr------------
# | C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s:11:10: error: CHECK: expected string not found in input
# | # CHECK: error: Assembly input parsing had errors, use -skip-unsupported-instructions=parse-failure to drop failing lines from the input.
# |          ^
# | <stdin>:1:1: note: scanning from here
# | 'skylake-avx512' is not a recognized processor for this target (ignoring processor)
# | ^
# | 
# | Input file: <stdin>
# | Check file: C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\test\tools\llvm-mca\bad-input.s
# | 
# | -dump-input=help explains the following input dump.
# | 
# | Input was:
# | <<<<<<
# |           1: 'skylake-avx512' is not a recognized processor for this target (ignoring processor) 
# | check:11     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
# |           2: 'skylake-avx512' is not a recognized processor for this target (ignoring processor) 
# | check:11     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1
```

The bots for which this have failed appear to be cross-compilers which don't have the native build target enabled.

It seems the safest immediate fix is to move/copy this test to x86/arm so that there is some coverage for this test for two common targets.

Comment (peterwaller-arm) [reply llvm/test/tools/llvm-mca/bad-input.s:5]: Proposed fix committed as afc10fc9b7ce3d23d9012f5a1496e849fe873ba2 for now. I'll monitor the bots.

Comment (peterwaller-arm) [reply llvm/test/tools/llvm-mca/bad-input.s:5]: Not quite there. Looks like we can't rely on the default mcpu for bad inputs, since this fails if we run on a CPU for which there is no scheduling information.

So I've gone for supplying mtriple and mcpu for this case, committed in 458d70674190c4d043d5dfd2e41aecddff5cdb69.

**Pull request merged**
